### ISSUE TITLE
[New Concept Docs]: `itertools` Module

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3089]
This issue describes how to implement the `itertools` module  **concept docs**.
You can find the related **concept exercise** issue [here](https://github.com/exercism/python/issues/3088)

If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.

## ✅   Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🎯   Goal

These concept docs are meant to teach an understanding/use of the `itertools` module  in Python. 

<br>

## 💡   Learning objectives

Learn more about `iteration` tools the Python Standard Library provides through the `itertools` module.

Build and understanding of and use the following functions  from the module, as well as practicing some of the [recipes](https://docs.python.org/3/library/itertools.html#itertools-recipes) included :

- At least one of the _infinite itertators_ `count()`,`cycle()`,, or `repeat()`
- `accumulate()`
- `product()`
- `chain()` & `chain.from_iterable()`
- `groupby()`
- `islice()`
- `zip_longest()` and the `zip() built-in`
- `permutations()`
- `combinations()`

<br>

## 🚫  Out of scope

<br>
<details>
<summary>Concepts & Subjects that are Out of Scope (<em>click to expand</em>)</summary>
<br>

- `classes`  & `class customization` beyond the use of the `itertools`  methods.
- `class-inheritance` beyond what is needed to customize `iteration` using `itertools`
- `comprehensions` beyond what is needed to work with `itertools`
- `comprehensions` in `lambdas`
- `coroutines` 
- `decorators` beyond what is needed to work with `itertools`
- `functions` and `higher-order functions`  beyond what might be needed to work with `itertools`
-  `functools` and related `map()`, `filter()` and `functools.reduce()`(_they have their own exercise which is a prerequisite to this one_)
- `generators` beyond what might be needed to work with `itertools` (_they have their own exercise which is a prerequisite to this one_)
- `lambdas`  beyond what might be needed to work with `itertools`
- using an `assignment expression` or "walrus" operator (`:=`) 
- class decorators 
- `enums`

</details>
<br>

## 🤔  Concepts

- `iteration`
- `iterators`
- `itertools`

<br>

## ↩️   Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

<details>
<summary>Exercise Prerequisites (<em>click to expand</em>)</summary>
<br>

- `basics`
- `booleans`
- `comparisons`
- `rich-comparisons`
- `dicts`
- `dict-methods`
- `functions`
- `functional tools`
- `generators`
- `higher-order functions`
- Identity methods `is` and `is not`
- `iteration`
- `lists`
- `list-methods`
- `loops`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

</details>

<br>

## 📚  Resources to refer to

<details>
<summary>Resources (<em>click to expand</em>)</b></summary>
<br>

- [Python Docs: iterator definition](https://docs.python.org/3/glossary.html#term-iterator) | [Python Docs: iterator types](https://docs.python.org/3/library/stdtypes.html#iterator-types) 
  -  [Python Functional Programming HOWTO: Iterators](https://docs.python.org/3/howto/functional.html#iterators)
  -  [Trey Hunner: The Iterator Protocol: How `for` Loops Work in Python](https://treyhunner.com/2016/12/python-iterator-protocol-how-for-loops-work/)
  -  [Dan Bader: Python Iterators - A Step-by-Step Introduction](https://dbader.org/blog/python-iterators)
  -   [Python Docs: sequence](https://docs.python.org/3/glossary.html#term-sequence)
- [Python Docs: `itertools` module](https://docs.python.org/3/library/itertools.html#module-itertools)
- [Python Docs:`itertools` Recipes](https://docs.python.org/3/library/itertools.html#itertools-recipes)
- [Python Module of the Week: `itertools`](https://pymotw.com/3/itertools/)
- [Florian Dahlitz:  Python Introduction to Itertools](https://florian-dahlitz.de/blog/introduction-to-itertools)
- [Real Python: Itertools in Python 3, By Example](https://docs.python.org/3/library/itertools.html#module-itertools)
- [Python Docs: eumulating container types](https://docs.python.org/3/reference/datamodel.html#emulating-container-types)
  - [Trey Hunner: How to make an `iterator` in Python](https://treyhunner.com/2018/06/how-to-make-an-iterator-in-python/)
  - [Python Patterns Guide: Gang of Four `iterator` Pattern](https://python-patterns.guide/gang-of-four/iterator/)

</details>
<br>

##  📁  Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b></summary>
<table>
<td>
<br>

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Concept `.meta/config.json` Entries
 
  For more information, see [Concept `.meta/config.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-metaconfigjson)

  * This file is likely already stubbed out.  Remember to add a concept **blurb** of less than 350 characters.  Please also add your GitHub username to the "authors" array, and any contributor GitHub usernames to the "contributors" array.

<br>
</td>
</table>
</details>

## 🎶  Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.** Where possible, please use REPL formatting, unless you are demonstrating pseudo code or a long code block.
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.
